### PR TITLE
feat: compseAndIssueCredential API

### LINKED ISSUE #131: composeAndIssueCredential API
Implement the [composeAndIssueCredential](https://w3c-ccg.github.io/vc-issuer-http-api/#/internal/composeAndIssueCredential)API.

### PR DESCRIPTION
**Pull request by rolsonquadras**
- Implement https://w3c-ccg.github.io/vc-issuer-http-api/#/internal/composeAndIssueCredential API
- BDD test

closes #131 

Signed-off-by: Rolson Quadras <rolson.quadras@securekey.com>

### CODE REVIEW

File: pkg/restapi/vc/operation/models.go [hunk 1]
```diff
@@ -61,6 +62,22 @@ type IssueCredentialOptions struct {
...
+	Types                   []string               `json:"types,omitempty"`
+	IssuanceDate            *time.Time             `json:"issuanceDate,omitempty"`
+	ExpirationDate          *time.Time             `json:"expirationDate,omitempty"`
+	Claims                  map[string]interface{} `json:"claims,omitempty"`
+	Evidence                verifiable.Evidence    `json:"evidence,omitempty"`
+	TermsOfUse              json.RawMessage        `json:"termsOfUse,omitempty"`
+	CredentialFormat        string                 `json:"credentialFormat,omitempty"`
+	ProofFormat             string                 `json:"proofFormat,omitempty"`
+	CredentialFormatOptions json.RawMessage        `json:"credentialFormatOptions,omitempty"`
+	ProofFormatOptions      json.RawMessage        `json:"proofFormatOptions,omitempty"`
```
Comment (llorllale) [inline pkg/restapi/vc/operation/models.go]: q: why are `ProofFormatOptions` and `CredentialFormatOptions` of type `json.RawMessage`, but `Claims` is a `map[string]interface{}`?

File: pkg/restapi/vc/operation/models.go [hunk 2]
```diff
@@ -61,6 +62,22 @@ type IssueCredentialOptions struct {
...
 
+// ComposeCredentialRequest for composing and issuing credential.
+type ComposeCredentialRequest struct {
+	Issuer                  string                 `json:"issuer,omitempty"`
+	Subject                 string                 `json:"subject,omitempty"`
+	Types                   []string               `json:"types,omitempty"`
+	IssuanceDate            *time.Time             `json:"issuanceDate,omitempty"`
+	ExpirationDate          *time.Time             `json:"expirationDate,omitempty"`
+	Claims                  map[string]interface{} `json:"claims,omitempty"`
+	Evidence                verifiable.Evidence    `json:"evidence,omitempty"`
```
Comment (llorllale) [inline pkg/restapi/vc/operation/models.go]: nitpick: I believe part of the point of a "compose" API is to decouple the credential format from the inputs. Therefore, `Evidence` should not be typed as `verifiable.Evidence`.

File: pkg/restapi/vc/operation/operations.go [hunk 3]
```diff
@@ -828,39 +830,145 @@ func (o *Operation) issueCredentialHandler(rw http.ResponseWriter, req *http.Req
...
-	publicKeyID, err := getPublicKeyID(didDoc)
+	rw.WriteHeader(http.StatusOK)
+	o.writeResponse(rw, signedVC)
+}
+
+func (o *Operation) composeAndIssueCredentialHandler(rw http.ResponseWriter, req *http.Request) {
+	// get the request
+	composeCredReq := ComposeCredentialRequest{}
+
+	err := json.NewDecoder(req.Body).Decode(&composeCredReq)
```
Comment (llorllale) [inline pkg/restapi/vc/operation/operations.go]: Any followups for the two query parameters, `templateReference` and `subjectReference`?

File: pkg/restapi/vc/operation/operations.go [hunk 4]
```diff
@@ -828,39 +830,145 @@ func (o *Operation) issueCredentialHandler(rw http.ResponseWriter, req *http.Req
...
 
-	rw.WriteHeader(http.StatusOK)
-	o.writeResponse(rw, signedVC)
+	// set subject
+	if composeCredReq.Claims == nil {
+		composeCredReq.Claims = make(map[string]interface{})
+	}
+
+	composeCredReq.Claims["subject"] = composeCredReq.Subject
+	credential.Subject = composeCredReq.Claims
```
Comment (llorllale) [inline pkg/restapi/vc/operation/operations.go]: These two lines are confusing. Why are we doing this?

File: pkg/restapi/vc/operation/operations.go (hunk 3 above)
Comment (rolsonquadras) [reply pkg/restapi/vc/operation/operations.go]: https://github.com/trustbloc/edge-service/issues/144

File: pkg/restapi/vc/operation/models.go (hunk 1 above)
Comment (rolsonquadras) [reply pkg/restapi/vc/operation/models.go]: Updated.

File: pkg/restapi/vc/operation/operations.go (hunk 4 above)
Comment (rolsonquadras) [reply pkg/restapi/vc/operation/operations.go]: Combining `subject` and `claims` into `credentialSubject`.

File: test/bdd/pkg/issuer/issuer_steps.go [hunk 5]
```diff
@@ -204,6 +247,61 @@ func (e *Steps) signCredential(didDocVar string) ([]byte, error) {
...
+	if err != nil {
+		return err
+	}
+
+	proof, ok := signedVCResp["proof"].(map[string]interface{})
+	if !ok {
+		return errors.New("unable to convert proof to a map")
+	}
+
+	if proof["type"] != "Ed25519Signature2018" {
```
Comment (troyronda) [inline test/bdd/pkg/issuer/issuer_steps.go]: TODO for the NIST P256 version?

Comment (rolsonquadras) [reply test/bdd/pkg/issuer/issuer_steps.go]: https://github.com/trustbloc/edge-service/issues/new

File: pkg/restapi/vc/operation/models.go (hunk 2 above)
Comment (rolsonquadras) [reply pkg/restapi/vc/operation/models.go]: Missed this. Makes sense, will add it in next PR.

**Pull request merged**